### PR TITLE
[JIT] add support for torch.jit.Final in python 3.6

### LINKED ISSUE #46964: torch.jit.Final does not work in Pytorch1.7 + Python3.6
## To Reproduce

```
# Dockerfile
FROM python:3.6-slim-buster
USER root
WORKDIR /root
RUN python3 -m pip install --upgrade pip
RUN python3 -m pip install torch==1.7.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
COPY a.py /
RUN python3 /a.py

# a.py:
import torch
from torch import nn

class A(nn.Module):
    dim: torch.jit.Final[int] = 4

    def __init__(self):
        super().__init__()
        self.l = nn.Linear(10, 10)

    def forward(self, x):
        return self.l(x)


class B(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = A()

    def forward(self, x):
        return self.a(x) + self.a.dim

b = B()
b = torch.jit.script(b)
```

Error:
```
Traceback (most recent call last):
  File "/a.py", line 24, in <module>
    b = torch.jit.script(b)
  File "/usr/local/lib/python3.6/site-packages/torch/jit/_script.py", line 898, in script
    obj, torch.jit._recursive.infer_methods_to_compile
  File "/usr/local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 352, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/usr/local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 410, in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
  File "/usr/local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 304, in create_methods_and_properties_from_stubs
    concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
RuntimeError: 
Module 'A' has no attribute 'dim' :
  File "/a.py", line 21
    def forward(self, x):
        return self.a(x) + self.a.dim
                           ~~~~~~~~~~ <--- HERE
```

It works if we change 3.6 to 3.7.

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47393 [JIT] add support for torch.jit.Final in python 3.6**

Python added support for typing.Final in 3.8, and there exists a backport for 3.7 and 3.6 in typing_extensions.Final

The 3.6 version of typing_extensions.Final doesn't support the `getattr(x, '__origin__')` to determine if an object is of the Final type while 3.7 does. This PR allows for the Final type of an object to be discoverable in 3.6

Differential Revision: [D24739402](https://our.internmc.facebook.com/intern/diff/D24739402)

### CODE REVIEW

File: torch/_jit_internal.py [hunk 1]
```diff
@@ -717,7 +717,7 @@ def is_rref(ann):
 
 def is_final(ann):
     return ann.__module__ in {'typing', 'typing_extensions'} and \
-        (getattr(ann, '__origin__', None) is Final)
+        (getattr(ann, '__origin__', None) is Final or isinstance(ann, type(Final)))
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:720]: Optional:

Can you say a bit more about why this change is needed (in PR description)? I am curious.

**Review by gmagogsfm — APPROVED** [review]

**Comment by SplitInfinity** [comment]
Do we need a test for this? I think the issue that reported this issue has a small test case we can use.

File: torch/_jit_internal.py (hunk 1 above)
Comment (Lilyjjo) [reply torch/_jit_internal.py:720]: added :)

File: test/jit/test_recursive_script.py [hunk 2]
```diff
@@ -159,7 +158,6 @@ def a(x):
             # Make sure that no entries are left over from the previous failure
             FileCheck().check_count("is being compiled", 2).run(str(e))
 
-    @unittest.skipIf(sys.version_info[:2] < (3, 7), "Class annotations are a thing in > 3.5, need to fix for < 3.7")
```
Comment (Lilyjjo) [inline test/jit/test_recursive_script.py:162]: I think this comment was out of date? Not totally sure, but this test now passes for python3.6 and python3.7 and checks the functionality of Final
